src/runtime-rs/crates/hypervisor/src/remote/inner.rs

```diff
@@ -0,0 +1,330 @@
+
+use super::protocols::{
+    hypervisor::{StartVMRequest, StopVMRequest},
+    hypervisor_ttrpc::HypervisorClient,
+};
+
+const REMOTE_SCHEME: &str = "remote";
+const DEFAULT_MIN_TIMEOUT: i32 = time::Duration::from_secs(60).as_millis() as i32;
+
+#[derive(Debug)]
+pub struct RemoteInner {
+    /// sandbox id
+    pub(crate) id: String,
+    /// hypervisor config
+    pub(crate) config: HypervisorConfig,
+    /// agent socket path
+    pub(crate) agent_socket_path: String,
+    /// sandbox annotations
+    pub(crate) annotations: HashMap<String, String>,
+    /// netns path
+    pub(crate) netns: Option<String>,
+
+    exit_notify: Option<mpsc::Sender<i32>>,
+    exit_waiter: Mutex<(mpsc::Receiver<i32>, i32)>,
+}
+
+impl RemoteInner {
+    pub fn new() -> Self {
+        let (exit_notify, exit_waiter) = mpsc::channel(1);
+
+        Self {
+            id: "".to_string(),
+            config: HypervisorConfig::default(),
+            agent_socket_path: "".to_string(),
+            annotations: HashMap::new(),
+            netns: None,
+
+            exit_notify: Some(exit_notify),
+            exit_waiter: Mutex::new((exit_waiter, 0)),
+        }
+    }
+
+    pub(crate) async fn prepare_vm(&mut self, id: &str, netns: Option<String>) -> Result<()> {
+        info!(sl!(), "Preparing REMOTE VM");
+        self.id = id.to_string();
+        let c = Client::connect(&format!("unix://{}", &self.config.remote_hypervisor_socket))?;
+        let client = HypervisorClient::new(c.clone());
+
+        if let Some(netns_path) = &netns {
+            debug!(sl!(), "set netns for vmm master {:?}", &netns_path);
+            let netns_fd = std::fs::File::open(netns_path);
+            if netns_fd.is_err() {
+                error!(sl!(), "failed to open netns file {:?}", netns_path);
+            } else {
+                info!(sl!(), "set netns for vmm master {:?}", &netns_path);
+            }
+        }
+
+        let ctx = context::with_timeout(10_0000_0000);
+        let req = CreateVMRequest {
+            id: id.to_string(),
+            annotations: self.annotations.clone(),
+            networkNamespacePath: netns.clone().unwrap_or_default(),
+            ..Default::default()
+        };
+        info!(sl!(), "Preparing REMOTE VM req: {:?}", req.clone());
+        let resp = client.create_vm(ctx, &req).await.unwrap();
+        info!(sl!(), "Preparing REMOTE VM resp: {:?}", resp.clone());
+        self.agent_socket_path = resp.agentSocketPath;
+        self.netns = netns;
+        Ok(())
+    }
+
+    pub(crate) async fn start_vm(&mut self, timeout: i32) -> Result<()> {
+        info!(sl!(), "Starting REMOTE VM");
+
+        let mut min_timeout = DEFAULT_MIN_TIMEOUT;
+        if self.config.remote_hypervisor_timeout > 0 {
+            min_timeout = self.config.remote_hypervisor_timeout.min(timeout);
+        }
+        let timeout = min_timeout;
+
+        let c = Client::connect(&format!("unix://{}", &self.config.remote_hypervisor_socket))?;
```

Comment:
Could you please abstract the Client::connect(xxx) into a dedicated method to handle such similar works ?

Reply:
I have added `new_ttrpc_client` method for `RemoteInner` to avoid them.